corehq/form_processor/steps.py

```diff
@@ -0,0 +1,59 @@
+import re
+from abc import ABC, abstractmethod
+
+from corehq.apps.data_vault import new_vault_entry
+
+
+class FormProcessingStep(ABC):
+    @abstractmethod
+    def __call__(self, context):
+        """
+        Process the form context object. Return an instance of
+        FormProcessingResult to terminate the processing or None
+        To allow processing to continue to the next step
+        :param context: form context
+        """
+        pass
+
+
+class VaultPatternExtractor(FormProcessingStep):
+    def __init__(self, patterns, xmlns_whitelist=None):
+        self._patterns = patterns
```

Comment:
these should be compiled to avoid having to compile them for every form

Reply:
good point
https://github.com/dimagi/commcare-hq/pull/28161/commits/f3eac0836e0e064f4c32b035f49e62cd1b0f63b5